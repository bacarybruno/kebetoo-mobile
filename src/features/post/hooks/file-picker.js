import { useCallback, useState, useEffect } from 'react'
import dayjs from 'dayjs'
import RNFetchBlob from 'rn-fetch-blob'
import { useNavigation } from '@react-navigation/native'

import { api } from '@app/shared/services'
import { getExtension, getMimeType, isVideo } from '@app/shared/helpers/file'
import { CameraRollPicker } from '@app/shared/components'
import routes from '@app/navigation/routes'

export const constructFileName = (time, prefix = 'IMG', extension, duration) => {
  let fileName = `${prefix}-${time}`
  if (duration) {
    fileName = `${fileName}-${duration}`
  }
  fileName = `${fileName}.${extension}`
  return fileName
}

const useFilePicker = (uri) => {
  const [file, setFile] = useState(null)
  const { navigate } = useNavigation()

  const pickImage = useCallback(async () => {
    const fileData = await new Promise((resolve) => {
      navigate(routes.CAMERA_ROLL_PICKER, {
        assetType: CameraRollPicker.AssetTypes.Photos,
        onSelectedItem: (item) => {
          if (item) resolve(item)
          resolve(false)
        },
      })
    })
    if (fileData) setFile(fileData)
  }, [navigate])

  const pickVideo = useCallback(async () => {
    const fileData = await new Promise((resolve) => {
      navigate(routes.CAMERA_ROLL_PICKER, {
        assetType: CameraRollPicker.AssetTypes.Videos,
        onSelectedItem: (item) => {
          resolve(item || false)
        },
      })
    })
    if (fileData) setFile(fileData)
  }, [navigate])

  useEffect(() => {
    if (uri) {
      setFile({
        uri: uri.startsWith('file://') ? uri : `file://${uri}`,
        type: getMimeType(uri),
      })
    }
  }, [uri])

  const reset = useCallback(async () => {
    if (file) {
      try {
        if (!isVideo(file.uri)) {
          await RNFetchBlob.fs.unlink(file.uri)
        }
      } finally {
        setFile(null)
      }
    }
  }, [file])

  const savePost = useCallback(async (author, content, repost) => {
    // TODO: check if it's necessary to have unique file names
    const time = dayjs().format('YYYYMMDD')
    const fileUri = file.uri.replace('file://', '')
    let response
    if (isVideo(file.uri)) {
      response = await api.posts.createVideo({
        author,
        content,
        video: {
          uri: fileUri,
          mimeType: file.type,
          name: constructFileName(time, 'VID', getExtension(fileUri), file.duration),
        },
        repost,
      })
    } else {
      response = await api.posts.createImage({
        author,
        content,
        image: {
          uri: fileUri,
          mimeType: file.type,
          name: constructFileName(time, 'IMG', getExtension(fileUri)),
        },
        repost,
      })
    }
    await reset()
    return response
  }, [file, reset])

  const saveFeedback = useCallback(async (author, content, repost) => {
    // TODO: check if it's necessary to have unique file names
    const time = dayjs().format('YYYYMMDD')
    const fileUri = file.uri.replace('file://', '')
    const response = await api.feedbacks.createImage({
      author,
      content,
      image: {
        uri: fileUri,
        mimeType: file.type,
        name: constructFileName(time, 'IMG', getExtension(fileUri)),
      },
      repost,
    })
    await reset()
    return response
  }, [file, reset])

  return {
    pickImage,
    pickVideo,
    hasFile: file !== null,
    type: isVideo(file?.uri) ? 'video' : 'image',
    file,
    reset,
    savePost,
    saveFeedback,
  }
}

export default useFilePicker
