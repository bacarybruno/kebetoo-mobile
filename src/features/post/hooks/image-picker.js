import { useCallback, useState, useEffect } from 'react'
import ImagePicker from 'react-native-image-picker'
import dayjs from 'dayjs'
import RNFetchBlob from 'rn-fetch-blob'

import { api } from '@app/shared/services'
import { getMimeType } from '@app/shared/helpers/file'

const PICTURE_CONFIG = Object.freeze({
  mediaType: 'photo',
  noData: true,
  allowsEditing: true,
  quality: 0.8,
  maxHeight: 800,
})
export const constructFileName = (time) => `IMG-${time}`

const useFilePicker = (uri) => {
  const [file, setFile] = useState(null)

  const pickImage = useCallback(async () => {
    const fileData = await new Promise((resolve) => {
      ImagePicker.launchImageLibrary(PICTURE_CONFIG, (res) => {
        if (res.uri) resolve(res)
        else resolve(false)
      })
    })
    if (fileData) setFile(fileData)
  }, [])

  useEffect(() => {
    if (uri) {
      setFile({
        uri: uri.startsWith('file://') ? uri : `file://${uri}`,
        path: uri,
        type: getMimeType(uri),
      })
    }
  }, [uri])

  const reset = useCallback(async () => {
    if (file) {
      try {
        await RNFetchBlob.fs.unlink(file.path)
      } finally {
        setFile(null)
      }
    }
  }, [file])

  const savePost = useCallback(async (author, content, repost) => {
    const time = dayjs().format('YYYYMMDD')
    const response = await api.posts.createImage({
      author,
      content,
      image: {
        uri: file.uri,
        mimeType: file.type,
        name: constructFileName(time),
      },
      repost,
    })
    await reset()
    return response
  }, [file, reset])

  return {
    pickImage,
    hasFile: file !== null,
    file,
    reset,
    savePost,
  }
}

export default useFilePicker
