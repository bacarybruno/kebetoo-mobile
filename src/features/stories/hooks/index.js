export { default as useDarkBackground } from './dark-background';
export { default as useComments } from './comments';
