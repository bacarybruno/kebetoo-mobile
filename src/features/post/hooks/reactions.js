/* eslint-disable radix */
import React, { useCallback, useState, useEffect } from 'react'
import { useNavigation } from '@react-navigation/native'
import { useActionSheet } from '@expo/react-native-action-sheet'
import Ionicon from 'react-native-vector-icons/Ionicons'

import colors, { rgbaToHex } from '@app/theme/colors'
import routes from '@app/navigation/routes'
import { strings } from '@app/config'
import * as api from '@app/shared/services/http'
import { useAnalytics } from '@app/shared/hooks'

import { actionTypes } from '../containers/create'

export const REACTION_TYPES = {
  LIKE: 'like',
  DISLIKE: 'dislike',
  COMMENT: 'comment',
  LOVE: 'love',
  SHARE: 'share',
}

export const bottomSheetItems = [{
  title: strings.reactions.share_now,
  icon: 'ios-share-alt',
}, {
  title: strings.reactions.write_post,
  icon: 'ios-share',
}, {
  title: strings.general.cancel,
  icon: 'md-close',
}]

const createBottomSheetIcon = (item) => (
  <Ionicon name={item.icon} size={24} color={colors.textPrimary} />
)

const countReactions = (post, type) => (
  post.reactions.filter((r) => r.type === type).length
)

const useReactions = ({
  post: givenPost, author, comments, onComment,
}) => {
  const [post, setPost] = useState(givenPost)

  useEffect(() => {
    setPost({ ...givenPost })
  }, [givenPost])

  const { showActionSheetWithOptions } = useActionSheet()
  const { navigate } = useNavigation()
  const { trackSelectPost } = useAnalytics()

  const findUserReaction = useCallback((reaction) => reaction.author === author, [author])

  const userReaction = post.reactions.find(findUserReaction) || {}

  const createReaction = useCallback(async (type) => {
    // create random negative id
    const optimisticId = parseInt(Math.random() * -1000000)

    // display a reaction with expected value and a fake id
    const optimisticPost = {
      ...post,
      reactions: [
        ...post.reactions,
        { id: optimisticId, type, author },
      ],
    }
    setPost({ ...optimisticPost })

    // create the reaction on the backend
    api.createReaction(type, post.id, author)
      .then((res) => {
        // replace the fake reaction created with the new one
        const result = res
        result.post = result.post.id
        result.author = result.author.id
        optimisticPost.reactions = [
          ...optimisticPost.reactions.filter((reaction) => reaction.id !== optimisticId),
          result,
        ]
        // update post to trigger re-render
        setPost({ ...optimisticPost })
      }).catch(() => {
        // rollback the operation
        optimisticPost.reactions = optimisticPost.reactions.filter((r) => r.id !== optimisticId)
        setPost({ ...optimisticPost })
      })
  }, [author, post])

  const deleteReaction = useCallback(async (reactionId) => {
    // don't do anything if reactionId is not a valid ID
    if (parseInt(reactionId) < 0) return

    // remove reaction from state
    const optimisticPost = { ...post }
    const deletedReaction = optimisticPost.reactions.find((r) => r.id === reactionId)
    optimisticPost.reactions = optimisticPost.reactions.filter((r) => r.id !== reactionId)
    setPost({ ...optimisticPost })

    // then remove it from the backend
    api.deleteReaction(reactionId).catch(() => {
      // operation failed => rollback the operation
      optimisticPost.reactions = [
        ...optimisticPost.reactions,
        deletedReaction,
      ]
      setPost({ ...optimisticPost })
    })
  }, [post])

  const editReaction = useCallback(async (type, reactionId) => {
    // don't do anything if reactionId is not a valid ID
    if (parseInt(reactionId) < 0) return

    // find the reaction to edit and change the type
    const optimisticPost = { ...post }
    const editedReaction = optimisticPost.reactions.find((r) => r.id === reactionId)
    const editedReactionBaseType = editedReaction.type
    editedReaction.type = type

    // add the reaction in the state to trigger re-render
    optimisticPost.reactions = [
      ...optimisticPost.reactions.filter((r) => r.id !== reactionId),
      editedReaction,
    ]
    setPost({ ...optimisticPost })

    // then edit it from the backend
    api.editReaction(reactionId, type).catch(() => {
      // operation failed => rollback the operation
      editedReaction.type = editedReactionBaseType
      optimisticPost.reactions = [
        ...optimisticPost.reactions.filter((r) => r.id !== reactionId),
        editedReaction,
      ]
      setPost({ ...optimisticPost })
    })
  }, [post])

  const handlePostReaction = useCallback(async (type) => {
    const reaction = post.reactions.find(findUserReaction)
    if (reaction === undefined) {
      await createReaction(type)
    } else if (reaction.type === type) {
      await deleteReaction(reaction.id)
    } else {
      await editReaction(type, reaction.id)
    }
  }, [post.reactions, findUserReaction, createReaction, deleteReaction, editReaction])

  const handlePostShare = useCallback(() => {
    if (post.author.id === author) return
    // if ((post.repost?.author === author) === true) return
    const repostId = post.repost?.id || post.id
    const cancelButtonIndex = 2
    showActionSheetWithOptions({
      options: bottomSheetItems.map((item) => item.title),
      icons: bottomSheetItems.map(createBottomSheetIcon),
      cancelButtonIndex,
      title: strings.general.share,
      textStyle: { color: colors.textPrimary },
      titleTextStyle: { color: colors.textSecondary },
      containerStyle: { backgroundColor: rgbaToHex(colors.backgroundSecondary) },
    }, async (index) => {
      if (index === 0) {
        await api.createPost({ author, repost: repostId })
        navigate(routes.MANAGE_POSTS)
      } else if (index === 1) {
        navigate(routes.CREATE_POST, {
          action: actionTypes.SHARE,
          post: repostId,
        })
      }
    })
  }, [author, navigate, post, showActionSheetWithOptions])

  const onReaction = useCallback(async (type) => {
    switch (type) {
      case REACTION_TYPES.LIKE:
      case REACTION_TYPES.DISLIKE:
        handlePostReaction(type)
        break
      case REACTION_TYPES.COMMENT:
        if (onComment) {
          onComment()
        } else {
          navigate(routes.COMMENTS, { post })
          trackSelectPost(post.id)
        }
        break
      case REACTION_TYPES.SHARE:
        handlePostShare()
        break
      default: break
    }
    return null
  }, [handlePostReaction, onComment, navigate, post, handlePostShare, trackSelectPost])

  return {
    onReaction,
    userReactionType: userReaction.type,
    count: {
      likes: countReactions(post, REACTION_TYPES.LIKE),
      dislikes: countReactions(post, REACTION_TYPES.DISLIKE),
      comments: (comments || post.comments).length,
      shares: post.reposts?.length ?? 0,
    },
  }
}

export default useReactions
