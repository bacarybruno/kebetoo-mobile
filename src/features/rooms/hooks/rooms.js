import { useCallback, useEffect, useState } from 'react'
import database from '@react-native-firebase/database'

import { useUser } from '@app/shared/hooks'

const roomsPath = '/rooms'
const messagesPath = '/messages'
const onlinePath = '/online'

const systemMessages = {
  ROOM_CREATED: 'SYSTEM_ROOM_CREATED',
}

const useRooms = (roomId) => {
  const { profile } = useUser()
  const [roomsRef] = useState(database().ref(roomsPath))
  const [messagesRef] = useState(database().ref(messagesPath))
  const [onlineRef] = useState(database().ref(`${onlinePath}/${profile.uid}`))
  const [rooms, setRooms] = useState([])
  const [messages, setMessages] = useState([])
  const [discoverRooms, setDiscoverRooms] = useState([])
  const [onlineCount, setOnlineCount] = useState(0)

  useEffect(() => {
    if (roomId) {
      // get last 1000 messages
      messagesRef
        .child(roomId)
        .orderByChild('createdAt')
        .limitToLast(1000)
        .once('value', (snapshot) => {
          const values = []
          snapshot.forEach((child) => {
            values.unshift({ ...child.val(), received: true })
          })
          setMessages(values)
        })
      // appemd a message everytime a new one is created
      messagesRef
        .child(roomId)
        .orderByChild('createdAt')
        .startAt(Date.now())
        .on('child_added', (snapshot) => {
          const snapshotVal = {
            ...snapshot.val(),
            sent: true,
          }
          setMessages((oldState) => [snapshotVal, ...oldState])
        })
      // mark user as online in the room
      roomsRef
        .child(`/${roomId}/online/${profile.uid}`)
        .set(true)
      // mark user as offline on disconnect
      roomsRef
        .child(`/${roomId}/online/${profile.uid}`)
        .onDisconnect()
        .remove()
      // get the number of online users in the current room
      roomsRef
        .child(`/${roomId}/online`)
        .on('value', (snapshot) => {
          if (snapshot.val()) {
            setOnlineCount(Object.keys(snapshot.val()).length)
          } else {
            setOnlineCount(0)
          }
        })
    }
  }, [roomId, profile.uid, messagesRef, roomsRef])

  useEffect(() => {
    onlineRef.set({
      status: true,
      date: Date.now(),
    })
    onlineRef
      .onDisconnect()
      .set({
        status: false,
        date: Date.now(),
      })
  }, [onlineRef, profile])

  useEffect(() => {
    try {
      roomsRef.on('value', (data) => {
        const values = data.val()
        if (!values) return
        const roomsVal = Object.keys(values).map((key) => ({ id: key, ...values[key] }))
        const newRooms = []
        const newDiscoverRooms = []
        roomsVal.forEach((room) => {
          if (room.members && room.members[profile.uid]) {
            newRooms.unshift(room)
          } else {
            newDiscoverRooms.unshift(room)
          }
        })
        setRooms(newRooms)
        setDiscoverRooms(newDiscoverRooms)
      })
    } catch (error) {
      console.log(error)
    }
  }, [roomsRef, profile.uid])

  const quitRoom = useCallback(async () => {
    await roomsRef.child(`/${roomId}/members/${profile.uid}`).ref.remove()
    const roomMembers = await roomsRef.child(`/${roomId}/members`).once('value')
    if (roomMembers.exists() === false) {
      // the deleted member was the last one
      // the room must be deleted
      // gg BoeD
      await roomsRef.child(roomId).remove()
      // also delete messages
      await messagesRef.child(roomId).remove()
    }
  }, [messagesRef, profile.uid, roomId, roomsRef])

  const createMessage = useCallback(async ({
    room, text, audio, ...other
  }) => {
    const newMessage = messagesRef.child(`/${room}`).push()
    const createdAt = Date.now()
    // create the message
    // must follow this format: https://github.com/FaridSafi/react-native-gifted-chat#message-object
    const user = {
      _id: profile.uid,
      name: profile.displayName,
      avatar: profile.photoURL,
    }
    const message = {
      _id: newMessage.key,
      text: text || null,
      audio: audio || null,
      createdAt,
      user,
      ...other,
    }
    await newMessage.set(message, () => {
      setMessages((oldState) => {
        const updatedState = [...oldState]
        for (let i = 0; i < updatedState.length; i += 1) {
          const pendingMessage = updatedState[i]
          // eslint-disable-next-line no-underscore-dangle
          if (pendingMessage._id === message._id) {
            pendingMessage.received = true
          }
        }
        return updatedState
      })
    })

    // set last message
    roomsRef.child(`/${room}/lastMessage`).set(message)

    const roomMembers = await roomsRef.child(`/${room}/members`).once('value')
    const isMember = roomMembers.hasChild(profile.uid)
    // add user in the room when he sent a message to that room
    if (!isMember) {
      roomMembers.ref.update({ [profile.uid]: Date.now() })
    }
  }, [messagesRef, profile, roomsRef])

  const createRoom = useCallback(async ({ name, theme }) => {
    const newRoom = roomsRef.push()
    const roomInfos = {
      name,
      theme,
      createdAt: new Date().toISOString(),
      author: {
        displayName: profile.displayName,
        uid: profile.uid,
        photoURL: profile.photoURL,
      },
      members: {
        [profile.uid]: Date.now(),
      },
      lastMessage: null,
      _id: newRoom.key,
    }
    await newRoom.set(roomInfos)
    await createMessage({
      text: systemMessages.ROOM_CREATED,
      system: true,
      room: newRoom.key,
    })
    return (await roomsRef.child(newRoom.key).once('value')).val()
  }, [roomsRef, profile, createMessage])

  return {
    rooms,
    messages,
    onlineCount,
    discoverRooms,
    quitRoom,
    createRoom,
    createMessage,
  }
}

export default useRooms
