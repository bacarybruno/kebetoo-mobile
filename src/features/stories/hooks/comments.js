import { useEffect, useCallback, useReducer } from 'react'

import { useUser, usePosts, useAudioRecorder } from '@app/shared/hooks'
import { api } from '@app/shared/services'
import reducer, { initialState, actionTypes } from '@app/features/comments/containers/reducer'

/**
 * Hackernews' hot sort
 * https://medium.com/hacking-and-gonzo/how-hacker-news-ranking-algorithm-works-1d9b0cf2c08d
 */
export const scoreComment = (votes, itemDate, gravity = 1.8) => {
  const hourAge = (Date.now() - itemDate.getTime()) / (1000 * 3600)
  return (votes - 1) / (hourAge + 2 ** gravity)
}

export const sortComments = (comments) => (
  comments
    .map((comment) => ({
      ...comment,
      score: scoreComment(
        comment.reactions.length + comment.replies.length,
        new Date(comment.createdAt),
      ),
    }))
    .sort((a, b) => {
      if (a.score < b.score) return 1
      if (b.score < a.score) return -1
      return 0
    })
)

const mapComments = (comment) => ({
  id: comment.id,
  author: {
    id: comment.author,
    displayName: ' ',
    photoURL: null,
  },
  reactions: [],
})

// TODO: paginate comments
const useComments = (story, commentInput, scrollView) => {
  const audioRecorder = useAudioRecorder()
  const initialComments = story.comments?.map(mapComments)
  if (initialComments) {
    initialState.comments = initialComments
  }

  const [state, dispatch] = useReducer(reducer, initialState)

  const {
    authors, comment, toReply, comments, replies, isLoading,
  } = state

  const { profile } = useUser()

  const { getRepostAuthors } = usePosts()

  useEffect(() => {
    const fetchComments = async () => {
      const result = await api.comments.getByStoryId(story.id)
      dispatch({ type: actionTypes.SET_COMMENTS, payload: sortComments(result) })
    }
    fetchComments()
  }, [story.id])

  useEffect(() => {
    const fetchRepostAuthors = async () => {
      const data = await getRepostAuthors(story)
      dispatch({ type: actionTypes.SET_AUTHORS, payload: data })
    }
    fetchRepostAuthors()
  }, [story, getRepostAuthors])

  const startLoading = useCallback(() => {
    dispatch({ type: actionTypes.START_LOADING })
  }, [])

  const endLoading = useCallback(() => {
    dispatch({ type: actionTypes.END_LOADING })
  }, [])

  const setLoading = useCallback((loading) => {
    if (loading) startLoading()
    else endLoading()
  }, [endLoading, startLoading])

  const onSend = useCallback(async () => {
    setLoading(true)

    try {
      let result = null
      let replyThread = null
      if (toReply) {
        // use the comment reply thread or the base comment
        replyThread = toReply.thread || toReply
      }

      if (audioRecorder.hasRecording) {
        result = await audioRecorder.saveStoryComment(story.id, profile.uid, replyThread)
      } else if (comment.length > 0) {
        result = await api.comments.create({
          author: profile.uid,
          content: comment,
          thread: replyThread ? replyThread.id : null,
          story: replyThread ? null : story.id,
        })
      }
      if (replyThread) {
        dispatch({
          type: actionTypes.ADD_REPLIES,
          payload: { threadId: replyThread.id, replies: result },
        })
      } else {
        dispatch({ type: actionTypes.ADD_COMMENT, payload: result })
        setTimeout(() => scrollView?.current?.scrollToEnd(), 200)
      }
    } finally {
      // cleanup
      commentInput.current?.clear()
      dispatch({ type: actionTypes.CLEAR_TO_REPLY })
      dispatch({ type: actionTypes.CLEAR_COMMENT })
      setLoading(false)
    }
  }, [setLoading, toReply, audioRecorder, comment, story.id, profile.uid, commentInput, scrollView])

  const onSetReply = useCallback((item) => {
    commentInput.current?.focus()
    dispatch({ type: actionTypes.SET_TO_REPLY, payload: item })
  }, [commentInput])

  const loadReplies = useCallback(async (thread) => {
    if (!replies[thread.id]) {
      dispatch({
        type: actionTypes.SET_REPLIES,
        payload: { threadId: thread.id, replies: thread.replies.map(mapComments) },
      })
    }
    const loadedReplies = await api.comments.getReplies(thread.id)
    dispatch({
      type: actionTypes.SET_REPLIES,
      payload: { threadId: thread.id, replies: loadedReplies },
    })
  }, [replies])

  const onComment = useCallback(() => commentInput.current?.focus(), [])

  const clearToReply = useCallback(() => {
    dispatch({ type: actionTypes.CLEAR_TO_REPLY })
  }, [])

  const setComment = useCallback((payload) => {
    dispatch({ type: actionTypes.SET_COMMENT, payload })
  }, [])

  return {
    onSend,
    onComment,
    onSetReply,
    setComment,
    loadReplies,
    clearToReply,
    audioRecorder,
    isLoading,
    authors,
    comments,
    toReply,
    replies,
    comment,
  }
}

export default useComments
