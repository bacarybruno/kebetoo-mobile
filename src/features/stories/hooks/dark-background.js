import { StatusBar } from 'react-native'
import { useContext } from 'react'
import changeNavigationBarColor from 'react-native-navigation-bar-color'

import { SafeAreaContext } from '@app/shared/contexts'
import iosColors from '@app/theme/ios-colors'
import { rgbaToHex } from '@app/theme/colors'
import { useAppColors } from '@app/shared/hooks'

const useDarkBackground = (isFocused, hide) => {
  const { resetAppBars } = useAppColors()

  const {
    resetStatusBars,
    updateTabBarTheme,
    updateTopSafeAreaColor,
    updateBottomSafeAreaColor,
  } = useContext(SafeAreaContext)

  if (hide) {
    updateTabBarTheme('hide')
    return
  }
  if (!isFocused) {
    resetStatusBars()
    resetAppBars()
    return
  }
  updateTopSafeAreaColor(iosColors.secondarySystemBackground.dark)
  updateBottomSafeAreaColor(iosColors.secondarySystemBackground.dark)
  changeNavigationBarColor(rgbaToHex(iosColors.secondarySystemBackground.dark))
  StatusBar.setBackgroundColor(iosColors.secondarySystemBackground.dark)
  updateTabBarTheme('dark')
  StatusBar.setBarStyle('light-content')
}

export default useDarkBackground
