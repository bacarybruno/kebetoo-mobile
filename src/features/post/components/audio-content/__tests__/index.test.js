import { render, fireEvent } from 'react-native-testing-library'

import { constructFileName } from '@app/shared/hooks/audio-recorder'
import setupTest from '@app/config/jest-setup'

import AudioContent from '../index'

const givenAudioContent = setupTest(AudioContent, render)({
  content: 'test audio',
  audioUrl: '/fake-audio.mp3',
  style: { elevation: 3 },
  onPress: jest.fn(),
  // date format: YYYYMMDD
  audioName: constructFileName(20200625, 10),
})

it('renders AudioContent', () => {
  const { wrapper } = givenAudioContent()
  expect(wrapper.toJSON()).toMatchSnapshot()
})

it('handles player press', () => {
  const { wrapper, props } = givenAudioContent()
  fireEvent.press(wrapper.getByTestId('player-wrapper'))
  expect(props.onPress).toBeCalledTimes(1)
})
