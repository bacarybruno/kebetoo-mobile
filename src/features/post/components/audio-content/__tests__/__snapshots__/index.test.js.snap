// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AudioContent 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "zIndex": 1,
        },
        false,
      ]
    }
  >
    <View
      accessibilityElementsHidden={false}
      importantForAccessibility="auto"
      onLayout={[Function]}
      pointerEvents="box-none"
      style={null}
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 0,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "rgb(255, 255, 255)",
              "borderBottomColor": "rgb(216, 216, 216)",
              "flex": 1,
              "shadowColor": "rgb(216, 216, 216)",
              "shadowOffset": Object {
                "height": 0.5,
                "width": 0,
              },
              "shadowOpacity": 0.85,
              "shadowRadius": 0,
            }
          }
        />
      </View>
      <View
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "maxHeight": undefined,
            "minHeight": undefined,
            "opacity": undefined,
            "transform": undefined,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "height": 0,
            }
          }
        />
        <View
          pointerEvents="box-none"
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "marginHorizontal": 16,
                "opacity": 1,
              }
            }
          >
            <Text
              accessibilityRole="header"
              numberOfLines={1}
              onLayout={[Function]}
              style={
                Object {
                  "color": "rgb(28, 28, 30)",
                  "fontSize": 17,
                  "fontWeight": "600",
                }
              }
            >
              jest
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      pointerEvents="box-none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "opacity": 1,
          }
        }
      />
      <View
        accessibilityElementsHidden={false}
        closing={false}
        gestureVelocityImpact={0.3}
        importantForAccessibility="auto"
        onClose={[Function]}
        onGestureBegin={[Function]}
        onGestureCanceled={[Function]}
        onGestureEnd={[Function]}
        onOpen={[Function]}
        onTransitionStart={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        transitionSpec={
          Object {
            "close": Object {
              "animation": "spring",
              "config": Object {
                "damping": 500,
                "mass": 3,
                "overshootClamping": true,
                "restDisplacementThreshold": 10,
                "restSpeedThreshold": 10,
                "stiffness": 1000,
              },
            },
            "open": Object {
              "animation": "spring",
              "config": Object {
                "damping": 500,
                "mass": 3,
                "overshootClamping": true,
                "restDisplacementThreshold": 10,
                "restSpeedThreshold": 10,
                "stiffness": 1000,
              },
            },
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            collapsable={false}
            forwardedRef={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            style={
              Object {
                "flex": 1,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "backgroundColor": "rgb(242, 242, 242)",
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 1,
                    "width": -1,
                  },
                  "shadowOpacity": 0.3,
                  "shadowRadius": 5,
                  "top": 0,
                  "width": 3,
                }
              }
            />
            <View
              style={
                Array [
                  Object {
                    "flex": 1,
                    "overflow": "hidden",
                  },
                  Array [
                    Object {
                      "backgroundColor": "rgb(242, 242, 242)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "flexDirection": "column-reverse",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "elevation": 3,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "backgroundColor": "transparent",
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontSize": 15,
                            "fontWeight": "400",
                            "letterSpacing": -0.24,
                            "lineHeight": 20,
                          },
                          Array [
                            Object {
                              "color": "rgba(0,0,0,0.87)",
                            },
                            undefined,
                            Object {
                              "marginBottom": 8,
                            },
                          ],
                        ]
                      }
                    >
                      test audio
                    </Text>
                    <View
                      style={
                        Array [
                          Object {
                            "flex": 1,
                            "height": 52,
                            "maxHeight": 52,
                          },
                          undefined,
                        ]
                      }
                    >
                      <View
                        accessible={true}
                        focusable={true}
                        onClick={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#F5F5F5",
                              "borderRadius": 5,
                              "flex": 1,
                              "overflow": "hidden",
                            },
                            Object {},
                            undefined,
                          ]
                        }
                        testID="player-wrapper"
                      >
                        <View
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "flexDirection": "row",
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "backgroundColor": "rgba(199, 199, 204, 255)",
                                "borderBottomLeftRadius": 5,
                                "borderTopLeftRadius": 5,
                                "height": "100%",
                                "opacity": 0.4,
                                "position": "absolute",
                                "width": "0%",
                              }
                            }
                            testID="progress"
                          />
                          <View
                            accessible={true}
                            focusable={true}
                            hitSlop={
                              Object {
                                "bottom": 10,
                                "left": 10,
                                "right": 10,
                                "top": 10,
                              }
                            }
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#F5F5F5",
                                "borderRadius": 1000,
                                "elevation": 2,
                                "height": 30,
                                "justifyContent": "center",
                                "marginLeft": 16,
                                "opacity": 1,
                                "shadowColor": "rgba(0, 0, 0, 255)",
                                "shadowOffset": Object {
                                  "x": 0,
                                  "y": 1,
                                },
                                "shadowOpacity": 0.2,
                                "shadowRadius": 1.41,
                                "width": 30,
                              }
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              style={
                                Object {
                                  "color": "#0B1B77",
                                  "fontFamily": "Ionicons",
                                  "fontSize": 22,
                                  "fontStyle": "normal",
                                  "fontWeight": "normal",
                                  "marginLeft": 3,
                                }
                              }
                              text="ionicon-ios-play"
                            />
                          </View>
                          <View
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                              }
                            }
                          >
                            <Image
                              source={
                                Object {
                                  "testUri": "../../../assets/images/waves.png",
                                }
                              }
                              style={
                                Object {
                                  "resizeMode": "contain",
                                  "width": "80%",
                                }
                              }
                            />
                          </View>
                          <Text
                            style={
                              Array [
                                Object {
                                  "backgroundColor": "transparent",
                                  "color": "#000000",
                                  "fontFamily": "System",
                                  "fontSize": 12,
                                  "fontWeight": "400",
                                  "letterSpacing": 0,
                                  "lineHeight": 16,
                                },
                                Array [
                                  Object {
                                    "color": "rgba(0,0,0,0.54)",
                                  },
                                  undefined,
                                  Array [
                                    Object {
                                      "bottom": 2,
                                      "position": "absolute",
                                      "right": 8,
                                    },
                                    undefined,
                                  ],
                                ],
                              ]
                            }
                          >
                            0:10
                          </Text>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
